xswap/xswap_backend.cpp

```diff
@@ -0,0 +1,194 @@
+#include <cstring>
+#include <Python.h>
+#include <random>
+
+#include "hash_table.h"
+
+int** py_list_to_edges(PyObject *py_list, int num_edges) {
+    int** edges_array;
+
+    edges_array = (int**)malloc(sizeof(int*) * num_edges);
+
+    for (int i = 0; i < num_edges; i++) {
+        edges_array[i] = (int*)malloc(sizeof(int) * 2);
+        PyObject* py_tuple = PyList_GetItem(py_list, i);
+        for (int j = 0; j < 2; j++) {
+            PyObject* temp = PyTuple_GetItem(py_tuple, j);
+            int value = (int)PyLong_AsLong(temp);
+            edges_array[i][j] = value;
+        }
+    }
+    return edges_array;
+}
+
+PyObject* edge_to_py_tuple(int *edge) {
+    PyObject* edge_tuple = PyTuple_New(2);
+    for (int j = 0; j < 2; j++) {
+        PyObject* node_id = PyLong_FromLong(edge[j]);
+        PyTuple_SET_ITEM(edge_tuple, j, node_id);
+    }
+    return edge_tuple;
+}
+
+PyObject* edges_to_py_list(int **edges, int num_edges) {
+    PyObject* py_list = PyList_New(num_edges);
+
+    for (int i = 0; i < num_edges; i++) {
+        PyObject* edge_tuple = edge_to_py_tuple(edges[i]);
+        PyList_SET_ITEM(py_list, i, edge_tuple);
+    }
+    return py_list;
+}
+
+PyObject* stats_to_py_dict(int num_swaps, int same_edge, int self_loop, int duplicate, int undir_duplicate, int excluded) {
+    PyObject* py_num_swaps = PyLong_FromLong(num_swaps);
+    PyObject* py_same_edge = PyLong_FromLong(same_edge);
+    PyObject* py_self_loop = PyLong_FromLong(self_loop);
+    PyObject* py_duplicate = PyLong_FromLong(duplicate);
+    PyObject* py_undir_duplicate = PyLong_FromLong(undir_duplicate);
+    PyObject* py_excluded = PyLong_FromLong(excluded);
+
+    PyObject* dict = PyDict_New();
+    int code;
+    code = PyDict_SetItemString(dict, "swap_attempts", py_num_swaps);
+    code = PyDict_SetItemString(dict, "same_edge", py_same_edge);
+    code = PyDict_SetItemString(dict, "self_loop", py_self_loop);
+    code = PyDict_SetItemString(dict, "duplicate", py_duplicate);
+    code = PyDict_SetItemString(dict, "undir_duplicate", py_undir_duplicate);
+    code = PyDict_SetItemString(dict, "excluded", py_excluded);
+    return dict;
+}
+
+static PyObject* _xswap(PyObject *self, PyObject *args) {
```

Comment:
Why this function is static? Is it required by Python when calling C++?
(Same question on the `static` functions on lines 179 and 184.)

Reply:
This function seems a little too heavy (over 100 lines). Does it make sense to split it into multiple ones?

Reply:
In reference to the use of `static`:
https://docs.python.org/3.7/extending/extending.html#providing-a-c-api-for-an-extension-module

> The details of visibility depend on the operating system [...] Portability therefore requires not to make any assumptions about symbol visibility. This means that all symbols in extension modules should be declared static, except for the module’s initialization function, in order to avoid name clashes with other extension modules

It may actually be better to have all Python-interacting functions defined in `xswap_backend.cpp` be static and to move non-Python-interacters to other files.